feat: define a max VNI value

SCedricThomas opened the pull request:
STORY-1341

In this PR, we are adding a max VNI value defined by env

SCedricThomas commented on idmanager/idmanager.go:
```diff
@@ -80,16 +83,17 @@ func (m *manager) Generate(ctx context.Context) (int, error) {
 
-	for i := 1; ; i++ {
+	// Searching for the first available ID until the maximum
+	for i := 1; i <= m.config.MaxVNI; i++ {
```
It is a bit weird to use the `MaxVNI` value here as we are referencing IDs here not VNIs.
How should I handle this? Is it alright to you?

EtienneM replied:
The naming is really bad, but this method is called from there:

https://github.com/Scalingo/sand/blob/bb2e01e837e178345c75e21d09a37f36c1d32e1c/network/create.go#L67

`vniGen` is defined there:

https://github.com/Scalingo/sand/blob/bb2e01e837e178345c75e21d09a37f36c1d32e1c/network/create.go#L59

From the code I sent, it's clear that it's a VNI generator. But for some reasons, the structure returned by the constructor is a `idmanager.Manager` (in a file named `vni_allocator.go... :facepalm:):

https://github.com/Scalingo/sand/blob/bb2e01e837e178345c75e21d09a37f36c1d32e1c/network/overlay/vni_allocator.go#L17-L19

Hence I understand your confusion about using the wording "VNI" here, but I actually don't think it's an issue. What do you think about adding a command to the `Generate` function to explicitly mention that it actually generates VNIs?

SCedricThomas commented on idmanager/idmanager_test.go:
```diff
@@ -2,94 +2,146 @@ package idmanager
+	"github.com/Scalingo/sand/store/storemock"
 )
 
 func TestManager_Generate(t *testing.T) {
```
I was not able to run the test locally and it was hanging forever:
```
panic: test timed out after 10m0s
	running tests:
		TestManager_Generate (10m0s)
		TestManager_Generate/when_no_item,_it_should_return_1,_then_2 (10m0s)

goroutine 195 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:2373 +0x385
created by time.goFunc
	/usr/local/go/src/time/sleep.go:215 +0x2d

goroutine 1 [chan receive, 10 minutes]:
testing.(*T).Run(0xc000147a00, {0xc7bc04?, 0x0?}, 0xcc38e8)
	/usr/local/go/src/testing/testing.go:1751 +0x3ab
testing.runTests.func1(0xc000147a00)
	/usr/local/go/src/testing/testing.go:2168 +0x37
testing.tRunner(0xc000147a00, 0xc000241c70)
	/usr/local/go/src/testing/testing.go:1690 +0xf4
testing.runTests(0xc00012b500, {0x12a1910, 0x1, 0x1}, {0x472410?, 0x47207a?, 0x12c3680?})
	/usr/local/go/src/testing/testing.go:2166 +0x43d
testing.(*M).Run(0xc000252780)
	/usr/local/go/src/testing/testing.go:2034 +0x64a
main.main()
	_testmain.go:45 +0x9b
``` 
I rewrote it and it seems to work now :/

SCedricThomas commented on config/config.go:
```diff
@@ -33,6 +33,8 @@ type Config struct {
 	EnableDockerPlugin   bool `envconfig:"ENABLE_DOCKER_PLUGIN"`
 	DockerPluginHttpPort int  `default:"9998"`
+
+	MaxVNI int `envconfig:"MAX_VNI" default:"999_999"`
```
Should we set this value by default to the specs defined value?

EtienneM replied:
Yes I think so

SCedricThomas commented:
Here are the unit tests passing:
```
?   	github.com/Scalingo/sand/api/httpresp	[no test files]
?   	github.com/Scalingo/sand/api/types	[no test files]
?   	github.com/Scalingo/sand/api/params	[no test files]
?   	github.com/Scalingo/sand/client/sand	[no test files]
?   	github.com/Scalingo/sand/cmd/sand-agent	[no test files]
?   	github.com/Scalingo/sand/endpoint	[no test files]
?   	github.com/Scalingo/sand/config	[no test files]
?   	github.com/Scalingo/sand/etcd	[no test files]
?   	github.com/Scalingo/sand/cmd/sand-agent-cli	[no test files]
?   	github.com/Scalingo/sand/integrations/docker	[no test files]
?   	github.com/Scalingo/sand/ipallocator	[no test files]
?   	github.com/Scalingo/sand/netlink	[no test files]
?   	github.com/Scalingo/sand/netnsbuilder	[no test files]
ok  	github.com/Scalingo/sand/idmanager	0.009s
ok  	github.com/Scalingo/sand/netutils	(cached)
?   	github.com/Scalingo/sand/network/netmanager	[no test files]
ok  	github.com/Scalingo/sand/network	(cached)
?   	github.com/Scalingo/sand/network/overlay/overlaymock	[no test files]
?   	github.com/Scalingo/sand/store/storemock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/client/sandmock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/endpointmock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/idmanagermock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/ipallocatormock	[no test files]
ok  	github.com/Scalingo/sand/network/overlay	(cached)
ok  	github.com/Scalingo/sand/store	(cached)
?   	github.com/Scalingo/sand/test/mocks/netlinkmock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/network/netmanagermock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/networkmock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/network/overlaymock	[no test files]
?   	github.com/Scalingo/sand/utils/tls	[no test files]
?   	github.com/Scalingo/sand/test/mocks/overlaymock	[no test files]
?   	github.com/Scalingo/sand/test/mocks/netnsbuildermock	[no test files]
ok  	github.com/Scalingo/sand/web	(cached)
```

SCedricThomas commented:
I will add a test about assigning the MaxVNI. It is currently not tested

EtienneM commented on idmanager/idmanager.go:
```diff
@@ -13,6 +13,8 @@ import (
 	"github.com/Scalingo/sand/store"
 )
 
+var ErrNoIDAvailable = errors.New("no new ID available")
```
suggestion: with what I wrote below, suggestion
```suggestion
var ErrNoVNIAvailable = errors.New("no new VNI available")
```

EtienneM replied:
I'm not sure I agree that we should talk about "ID" everywhere; I would rather change to talk about VNI everywhere. But that's up to you

SCedricThomas commented:
I added a suggestion commit  https://github.com/Scalingo/sand/pull/288/commits/dd440970757423320314750d8e9cadd29809693d to keep the idmanager generic without mentionning VNIs. I removed the config from the struct to directly define the max value authorized

EtienneM commented on idmanager/idmanager.go:
```diff
@@ -23,20 +24,20 @@ type Unlocker interface {
 
 type manager struct {
-	store  store.Store
-	config *config.Config
```
Oh the config is never used but to get this max value? nice change!
